File: my-folder/0017-letter-combinations-of-a-phone-number/solution.cpp
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        if(!digits.size())
            return {};
        vector<string>res;
        int n=digits.size();
        unordered_map<char,string>m;
        m['2']="abc";
        m['3']="def";
        m['4']="ghi";
        m['5']="jkl";
        m['6']="mno";
        m['7']="pqrs";
        m['8']="tuv";
        m['9']="wxyz";
        
        vector<string>pool;
        for(char c:digits)
            pool.push_back(m[c]);
        string local="";
        
        track(digits,pool,local,res);
        return res;
    }
    
    void track(string digits, vector<string>pool, string local, vector<string>&res)
    {
        if(local.size()==digits.size())
        {
            res.push_back(local);
            return; 
        }
        
        string searchSpace=pool[local.length()];
        
        for(char x:searchSpace)
        {
            local.push_back(x);
            track(digits,pool,local,res);
            local.pop_back();
        }
        
    }
};

File: my-folder/0017-letter-combinations-of-a-phone-number/solution.java
class Solution {
    static List<String> res;
    String arr[]={"0","1","abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"};
    public List<String> letterCombinations(String digits) {
    
        res=new ArrayList<>();
        int l=digits.length();
        if(l==0){
            return res;
        }
        // else if(l==1){
        //     res.add(arr[digits.charAt(0)-'0'].toCharArray());
        //     return res;
        // }
        else{
         recur("",digits);
         return res;
        }
    }

    void recur(String comb,String digits){
        if(digits.length()==0){
            res.add(comb);
        }else{
            String letters=arr[digits.charAt(0)-'0'];
            for(int j=0;j<letters.length();j++){
                recur(comb+letters.charAt(j),digits.substring(1));
            }
        }
    }
}
